File: 48_positional_encoding_timeseries/python/examples/time_series.py
#!/usr/bin/env python3
"""
Time Series Encoding Example

This example demonstrates temporal encodings specifically designed
for financial time series data, including calendar and market session features.
"""

import sys
import os
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import torch
import numpy as np
from datetime import datetime, timedelta
from positional_encoding import (
    TimeSeriesSinusoidalEncoding,
    CalendarEncoding,
    MarketSessionEncoding,
    MultiScaleTemporalEncoding,
)


def main():
    print("Time Series Encoding Example")
    print("=" * 50)

    d_model = 64
    batch_size = 2
    seq_len = 24  # 24 hours

    # 1. Time Series Sinusoidal Encoding
    print("\n1. Multi-Scale Sinusoidal Encoding")
    print("-" * 40)

    ts_encoding = TimeSeriesSinusoidalEncoding(d_model)
    x = torch.zeros(batch_size, seq_len, d_model)
    encoded = ts_encoding(x)

    print(f"Scales: 1h, 24h (daily), 168h (weekly), 720h (monthly)")
    print(f"Dimension: {d_model}")
    print(f"Input shape: {x.shape}")
    print(f"Output shape: {encoded.shape}")

    # 2. Calendar Encoding
    print("\n2. Calendar Encoding")
    print("-" * 40)

    calendar = CalendarEncoding(d_model)

    # Generate timestamps for a week
    base_time = datetime(2024, 1, 1, 0, 0)  # Monday
    # Create datetime list for a week
    dt_list = [base_time + timedelta(hours=i) for i in range(168)]

    # Extract discrete calendar features
    dayofweek = torch.tensor([dt.weekday() for dt in dt_list]).unsqueeze(0)
    month = torch.tensor([dt.month - 1 for dt in dt_list]).unsqueeze(0)  # 0-indexed
    quarter = torch.tensor([(dt.month - 1) // 3 for dt in dt_list]).unsqueeze(0)
    hour = torch.tensor([dt.hour for dt in dt_list]).unsqueeze(0)
    session = torch.zeros_like(hour)  # placeholder for session

    cal_encoded = calendar(dayofweek, month, quarter, hour, session)

    print(f"Calendar encoding dimension: {d_model}")
    print(f"Hour tensor shape: {hour.shape}")
    print(f"Encoded shape: {cal_encoded.shape}")

    # Show features for specific times
    print("\nCalendar features for different times:")
    for idx, name in [(0, "Monday 00:00"), (24, "Tuesday 00:00"),
                       (120, "Saturday 00:00"), (144, "Sunday 00:00")]:
        dt = base_time + timedelta(hours=idx)
        is_weekend = 1 if dt.weekday() >= 5 else 0
        print(f"  {name}: weekend={is_weekend}, encoded[0,:4]={cal_encoded[0, idx, :4].detach().numpy().round(3)}")

    # 3. Market Session Encoding
    print("\n3. Market Session Encoding")
    print("-" * 40)

    # Crypto market (24/7)
    print("\n3a. Cryptocurrency (24/7)")
    session_crypto = MarketSessionEncoding(d_model, market_type='crypto')

    # Generate hours for a day (MarketSessionEncoding for crypto expects hour tensor)
    hours_24 = torch.arange(24).unsqueeze(0)

    session_encoded = session_crypto(hours_24)
    print(f"Session encoding dimension: {d_model}")

    print("\nSession activity by hour (crypto):")
    for h in [0, 6, 12, 18]:
        # Determine session based on UTC hour
        if h < 8:
            sess_name = "Asia"
        elif h < 16:
            sess_name = "Europe"
        else:
            sess_name = "Americas"
        print(f"  {h:02d}:00 UTC - {sess_name} session")

    # Stock market
    print("\n3b. Stock Market (US)")
    session_stock = MarketSessionEncoding(d_model, market_type='stock')
    # For stock market, we need session and time_in_session tensors
    est_hours = (hours_24 - 5) % 24  # Convert UTC to EST

    # Compute session indices: 0=pre, 1=regular, 2=after, 3=closed
    session_idx = torch.full_like(hours_24, 3)  # default closed
    session_idx[(est_hours >= 4) & (est_hours < 9)] = 0   # pre-market
    session_idx[(est_hours >= 9) & (est_hours < 16)] = 1  # regular
    session_idx[(est_hours >= 16) & (est_hours < 20)] = 2 # after-hours

    # Time in session (minutes since session start)
    time_in_session = torch.zeros_like(hours_24)
    # For regular session (9:30-16:00), compute time in session
    regular_mask = session_idx == 1
    time_in_session[regular_mask] = ((est_hours[regular_mask] - 9) * 60).clamp(min=0, max=99)

    stock_encoded = session_stock(hours_24, session_idx, time_in_session)

    print("Session activity by hour (stock, EST = UTC-5):")
    for h in [9, 14, 17, 21]:  # UTC hours
        est_h = (h - 5) % 24
        # Match the session_idx logic above (integer hour boundaries)
        if 4 <= est_h < 9:
            sess_name = "Pre-market"
        elif 9 <= est_h < 16:
            sess_name = "Regular"
        elif 16 <= est_h < 20:
            sess_name = "After-hours"
        else:
            sess_name = "Closed"
        print(f"  {h:02d}:00 UTC ({est_h:.0f}:00 EST) - {sess_name}")

    # 4. Multi-Scale Temporal Encoding
    print("\n4. Multi-Scale Temporal Encoding")
    print("-" * 40)

    multi_scale = MultiScaleTemporalEncoding(d_model)
    # MultiScaleTemporalEncoding expects a dict of scale tensors
    ts_dict = {
        'minute': torch.zeros_like(hours_24),
        'hour': hours_24,
        'day': torch.tensor([dt.day - 1 for dt in dt_list[:24]]).unsqueeze(0),
        'week': torch.tensor([dt.weekday() for dt in dt_list[:24]]).unsqueeze(0),
        'month': torch.tensor([dt.month - 1 for dt in dt_list[:24]]).unsqueeze(0),
    }
    multi_encoded = multi_scale(ts_dict)

    print(f"Combined multi-scale encoding")
    print(f"Total dimension: {d_model}")
    print(f"Encoded shape: {multi_encoded.shape}")

    # 5. Practical Application
    print("\n5. Practical Application")
    print("-" * 40)

    print("\nRecommended encoding combinations by use case:")
    print()
    print("Intraday Trading (1-min to 1-hour):")
    print("  - TimeSeriesSinusoidal (multi-scale patterns)")
    print("  - MarketSession (session awareness)")
    print("  - RoPE (sequence position)")
    print()
    print("Swing Trading (4-hour to daily):")
    print("  - Calendar (day of week effects)")
    print("  - TimeSeriesSinusoidal (weekly/monthly patterns)")
    print("  - Sinusoidal (sequence position)")
    print()
    print("Long-term Investment (daily to monthly):")
    print("  - Calendar (month/quarter effects)")
    print("  - Learned (let model discover patterns)")
    print()
    print("Crypto 24/7:")
    print("  - MarketSession('crypto') - regional sessions")
    print("  - MultiScaleTemporalEncoding")
    print("  - RoPE (handles variable sequence lengths)")


if __name__ == "__main__":
    main()

File: 48_positional_encoding_timeseries/rust/examples/time_series.rs
//! Example: Time Series Encoding
//!
//! This example demonstrates temporal encodings specifically designed
//! for financial time series data, including calendar and market session features.

use positional_encoding::{
    CalendarEncoding, MarketSessionEncoding, MultiScaleTemporalEncoding,
    TimeSeriesSinusoidalEncoding, MarketType, PositionalEncoding,
};
use chrono::{DateTime, Utc, Datelike, Timelike};

fn main() {
    println!("Time Series Encoding Example");
    println!("============================\n");

    // 1. Time Series Sinusoidal Encoding
    println!("1. Multi-Scale Sinusoidal Encoding");
    println!("-----------------------------------");

    let ts_encoding = TimeSeriesSinusoidalEncoding::new(32);

    // Encode one week of hourly data
    let hourly_positions: Vec<usize> = (0..168).collect();
    let encoded = ts_encoding.encode(&hourly_positions);

    println!("Scales: 1h, 24h (daily), 168h (weekly), 720h (monthly)");
    println!("Dimension: {}", ts_encoding.dim());
    println!("One week of hourly data: {} positions", hourly_positions.len());
    println!("Output shape: {:?}", encoded.shape());

    // Show how different scales capture patterns
    println!("\nPattern capture:");
    println!("  Hour 0 (Monday 00:00): captures start of week");
    println!("  Hour 24 (Tuesday 00:00): 1 day elapsed, weekly pattern shifted");
    println!("  Hour 168 (Next Monday): weekly cycle complete");

    // 2. Calendar Encoding
    println!("\n2. Calendar Encoding");
    println!("--------------------");

    let calendar = CalendarEncoding::new();

    // Test timestamps for different scenarios
    let test_cases = vec![
        (1704067200, "2024-01-01 00:00 UTC (New Year, Monday)"),
        (1704110400, "2024-01-01 12:00 UTC (Monday noon)"),
        (1704499200, "2024-01-06 00:00 UTC (Saturday)"),
        (1706745600, "2024-02-01 00:00 UTC (Feb 1, Thursday)"),
    ];

    println!("Features: hour_sin/cos, dow_sin/cos, dom_sin/cos, month_sin/cos,");
    println!("          minute_sin/cos, week_sin/cos, is_weekend, month_start/end, quarter");
    println!("\nDimension: {}", calendar.dim());

    for (ts, desc) in &test_cases {
        let encoded = calendar.encode_timestamp(*ts);

        // Extract key features
        let is_weekend = encoded[12];
        let is_month_start = encoded[13];
        let is_month_end = encoded[14];
        let quarter = encoded[15];

        println!("\n{}", desc);
        println!("  Weekend: {:.0}, Month Start: {:.0}, Month End: {:.0}, Quarter: {:.2}",
            is_weekend, is_month_start, is_month_end, quarter);
    }

    // 3. Market Session Encoding
    println!("\n3. Market Session Encoding");
    println!("--------------------------");

    // Crypto market (24/7)
    println!("\n3a. Cryptocurrency (24/7)");
    let crypto_enc = MarketSessionEncoding::new(MarketType::Crypto);
    println!("Dimension: {}", crypto_enc.dim());
    println!("Features: is_asia, is_europe, is_americas, overlaps, hour_cycle, weekend, funding, midnight");

    let crypto_hours = vec![
        (1704070800, "01:00 UTC - Asia session"),
        (1704085200, "05:00 UTC - Asia session"),
        (1704099600, "10:00 UTC - Europe session"),
        (1704114000, "14:00 UTC - Europe/Americas overlap"),
        (1704128400, "18:00 UTC - Americas session"),
        (1704142800, "22:00 UTC - Americas session"),
    ];

    for (ts, desc) in &crypto_hours {
        let enc = crypto_enc.encode_timestamp(*ts);
        println!("{}: Asia={:.0}, Europe={:.0}, Americas={:.0}",
            desc, enc[0], enc[1], enc[2]);
    }

    // Stock market (9:30 AM - 4 PM EST)
    println!("\n3b. Stock Market (US)");
    let stock_enc = MarketSessionEncoding::new(MarketType::Stock);
    println!("Dimension: {}", stock_enc.dim());
    println!("Features: premarket, regular, afterhours, closed, session_progress, open/close_proximity, weekend");

    let stock_hours = vec![
        (1704099600, "10:00 UTC (5:00 EST) - Pre-market"),
        (1704114000, "14:00 UTC (9:00 EST) - Just before open"),
        (1704117600, "15:00 UTC (10:00 EST) - Regular session"),
        (1704139200, "21:00 UTC (16:00 EST) - Close"),
        (1704146400, "23:00 UTC (18:00 EST) - After hours"),
    ];

    for (ts, desc) in &stock_hours {
        let enc = stock_enc.encode_timestamp(*ts);
        println!("{}: Pre={:.0}, Reg={:.0}, After={:.0}, Closed={:.0}",
            desc, enc[0], enc[1], enc[2], enc[3]);
    }

    // Forex market
    println!("\n3c. Forex Market");
    let forex_enc = MarketSessionEncoding::new(MarketType::Forex);
    println!("Dimension: {}", forex_enc.dim());
    println!("Features: sydney, tokyo, london, ny, london_ny_overlap, tokyo_london_overlap, active_count");

    let forex_hours = vec![
        (1704070800, "01:00 UTC - Tokyo/Sydney"),
        (1704099600, "10:00 UTC - London"),
        (1704114000, "14:00 UTC - London/NY overlap (highest volume)"),
        (1704135600, "20:00 UTC - NY"),
    ];

    for (ts, desc) in &forex_hours {
        let enc = forex_enc.encode_timestamp(*ts);
        println!("{}: Sydney={:.0}, Tokyo={:.0}, London={:.0}, NY={:.0}",
            desc, enc[0], enc[1], enc[2], enc[3]);
    }

    // 4. Combined Multi-Scale Encoding
    println!("\n4. Multi-Scale Temporal Encoding");
    println!("---------------------------------");

    let multi_scale = MultiScaleTemporalEncoding::new(MarketType::Crypto);
    println!("Combines Calendar + Market Session encoding");
    println!("Total dimension: {}", multi_scale.dim());

    // Show how it encodes different market conditions
    let scenarios = vec![
        (1704117600, "Monday morning (Europe) - High activity expected"),
        (1704499200, "Saturday midnight - Low crypto activity"),
        (1704031200, "Month end Friday - Portfolio rebalancing"),
    ];

    for (ts, desc) in &scenarios {
        let enc = multi_scale.encode_timestamp(*ts);
        let dt = DateTime::from_timestamp(*ts, 0).unwrap();
        println!("\n{}", desc);
        println!("  Timestamp: {} ({})", ts, dt.format("%Y-%m-%d %H:%M UTC"));
        println!("  Encoding vector length: {}", enc.len());
        println!("  First 5 values: [{:.3}, {:.3}, {:.3}, {:.3}, {:.3}]",
            enc[0], enc[1], enc[2], enc[3], enc[4]);
    }

    // 5. Practical Application
    println!("\n5. Practical Application");
    println!("------------------------");

    println!("\nRecommended encoding combinations by use case:");
    println!();
    println!("Intraday Trading (1-min to 1-hour):");
    println!("  - TimeSeriesSinusoidal (multi-scale patterns)");
    println!("  - MarketSession (session awareness)");
    println!("  - RoPE (sequence position)");
    println!();
    println!("Swing Trading (4-hour to daily):");
    println!("  - Calendar (day of week effects)");
    println!("  - TimeSeriesSinusoidal (weekly/monthly patterns)");
    println!("  - Sinusoidal (sequence position)");
    println!();
    println!("Long-term Investment (daily to monthly):");
    println!("  - Calendar (month/quarter effects)");
    println!("  - Learned (let model discover patterns)");
    println!();
    println!("Crypto 24/7:");
    println!("  - MarketSession(Crypto) - regional sessions");
    println!("  - MultiScaleTemporalEncoding");
    println!("  - RoPE (handles variable sequence lengths)");

    println!("\nExample completed!");
}
